### PR TITLE
[WIP] Movement, Velocity, and Interpolation

### LINKED ISSUE #94: [Movement] Rubberbanding movement and Interpolation
**This issue began with work on orientation, which was not being updated correctly. The scope has since expanded to include all movement**

screenshot 48

### This will need
- [ ] Correct Position Update math (math in `physicsStep(...)`)
- [ ] (interp-v2, sending junk vals) Correctly send interpolated movement values in EntityUpdateCodec (`storeUnknownBinTree(...)`)
- [ ] Magic all of these values together
- [x] Correctly store ControlBits states (keypresses and releases) with keypresses time.
- [x] Do we need MotionStates (surface and physics params) for this to work? Probably.
- [x] Do we need TimeStates (record of when keys were pressed relative to server time) for this to work? Probably.
- [x] (interp-v2) Store InputStates to vector InputStateStorage
- [x] (interp-v2, I think 😜) Correct Velocity calculations -- Currently the velocity the client believes you to have, does not match with the server's expectation
- [x] Verify stops to server/client -- it appears as though the client doesn't *always* recognize that you've stopped moving (may relate to ControlBits)
- [x] Store Position updates to PosUpdates for interpolation
- [x] Calculate Interpolation values from PosUpdates (return BinTree)
- [x] Does EntityUpdateCodec send m_update_pos_and_cam every time? Or only for special updates?
- [X] Correct Entity Rotation upon spawn (Fixed by #104)
- [x] Correct Entity Rotation during movement in EntityUpdateCodec (addressed by #166)
- [x] Verify Speed calculations
- [x] Verify Drag calculations
- [x] Verify Gravity calculations
- [x] Client does not recognize ABS_TIME, though it is sent by the server

### Bonus
- Send Animations
- Send NetFX
- Correctly calculate jumping (this is going to be a whole other set of algos) 

### Useful testing commands (require Access Level 9)
- `/controldebug 1` - outputs to client console and shows markers in the world with position deltas
- `/entdebugclient 1` - displays GUI window that shows active data, and displays reticle on entity
- `/moveto {x} {y} {z}` - moves entity to position
- `/postest` - Outputs server values for position. Only available on my broxen/segs/interp-v2 branch
- `/interp` - Runs an interpolation test. Only available on my broxen/segs/interp-v2 branch

### SEGS version
78c7fd51e112b40b7db82527751a9686d2a0d87b  and later

### Tracking
Progress taking place in https://github.com/broxen/Segs/tree/movement

### PR DESCRIPTION
**Pull request by broxen**
Closes #94 

THIS IS IN PROGRESS AND SUBMITTED FOR REVIEW ONLY.

I'm putting this out there, because it's a much larger undertaking than I realized initially, and I'm hoping some feedback will be encouraging.

Feedback is greatly appreciated. The code is definitely a mess, so forgive me that.

Currently this contains the following changes:
- Remove temporary physics and update code from #241 and it's brethren.
- Add Velocity calculations
- Add PosUpdates during Entity load, and on Update
- Fixed rotation check in update_rot()
- Correctly handle keypress_release
- Separated InputState and InputStateStorage from InputStateEvent
- Moving toggles AFK
- Added some helper functions for interp testing
- In general movement is better, and movement of other entities is very smooth and mostly accurate position-wise.
- Added seqbits enum for future seq bits =D
- Code cleanup

TODO:
- Oh soooo very much...
- Keypress states still aren't fully saved with each state
- Keypress times aren't fully saved
- There's still a small amount of stuttering when moving
- Players move vertically, instead of staying on the ground. Gravity? Or, more likely, I need to ignore the y axis prior to using direction to update position.
- Interpolation crashes the client? Not sure why, something in bintree.
- Changed PosUpdates for bintree conversion to vector from array size 9. It might need to be a fixed size, and this could relate to the crashes when toggling interpolation
- TimeStates should probably be a part of InputStateStorage
- Movement flags and sequences should probably be saved along the way, but this may be beyond my original scope (eliminate rubberbanding)

### CODE REVIEW

File: Projects/CoX/Servers/MapServer/Events/InputState.h [hunk 1]
```diff
@@ -1,29 +0,0 @@
-/*
```
Comment (dracc) [inline Projects/CoX/Servers/MapServer/Events/InputState.h]: `git mv` instead of removing and re-creating?

Comment (broxen) [reply Projects/CoX/Servers/MapServer/Events/InputState.h]: From the git faq:
> Git has a rename command git mv, but that is just a convenience. The effect is indistinguishable from removing the file and adding another with different name and the same content

Comment (dracc) [reply Projects/CoX/Servers/MapServer/Events/InputState.h]: Oh. Til, thanks!

**Comment by broxen** [comment]
Will jump back on this after wrapping up #495

File: Projects/CoX/Servers/MapServer/EntityUpdateCodec.cpp [hunk 2]
```diff
@@ -110,44 +115,60 @@ bool storePosition(const Entity &src,BitStream &bs)
...
-bool update_rot(const Entity &/*src*/, int axis ) /* returns true if given axis needs updating */
+bool update_rot(const Entity &src, int axis ) /* returns true if given axis needs updating */
 {
-    int axis_tmp = axis; // Used to quell tautological comparison warning
-    if(axis==axis_tmp)   // FixMe: var compared against same var.
+    if(src.m_prev_state == nullptr)
         return true;
-    return false;
+
+    if(src.m_prev_state->m_orientation_pyr[axis] == src.m_cur_state->m_orientation_pyr[axis])
```
Comment (broxen) [inline Projects/CoX/Servers/MapServer/EntityUpdateCodec.cpp]: After running idle for several minutes, the server will segfault on this line.

Maybe it's due to m_prev_state being a pointer?

Comment (broxen) [reply Projects/CoX/Servers/MapServer/EntityUpdateCodec.cpp]: This should be resolved with the latest refactoring 32852f1

**Comment by broxen** [comment]
Closing this until it's closer to completion and ready for actual testing.

**Pull request closed without merging**